source/fundamentals/monitoring.txt

```diff
@@ -0,0 +1,409 @@
+instance. 
+
+An event is any action that happens in a running program. The driver includes functionality
+for listening to a subset of the events that occur when the driver is running.
+
+A listener is a class that performs some action when certain events occur. All
+listener classes extend the ``EventListener`` interface. A listener's
+API defines the events it can respond to.
+
+Each method of a listener class represents a response to a certain event. Each
+method receives one argument: an object representing the event the method
+responds to. All event objects are a subclass of the ``EventObject`` class.  
+
+The MongoDB Java driver organizes the events it defines into three categories:
+
+- Command Events
+- Server Discovery and Monitoring Events
+- Connection Pool Events
+
+The following sections show how to monitor each event category.
+
+For a full list of the events you can monitor,
+:java-docs:`see the event package of the MongoDB Java Driver <apidocs/mongodb-driver-core/com/mongodb/event/package-summary.html>`.
+
+For more information on the classes and interfaces mentioned in this section,
+see the following API documentation:
+
+- `EventListener <https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/EventListener.html>`__
+- `EventObject <https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/EventObject.html>`__
+
+Command Events
+~~~~~~~~~~~~~~
+
+A command event is an event related to a MongoDB database command. Some
+examples of database commands that produce command events are ``find``,
+``insert``, ``delete``, and ``count``. 
+
+To monitor command events, write a class that implements the
+``CommandListener`` interface and register an instance of that class with your
+``MongoClient`` instance.
+
+For more information on MongoDB database commands, see the
+:manual:`MongoDB manual entry on database commands </reference/command/>`.
+
+.. note:: Redacted Output
+
+   As a security measure, the driver is unable to monitor some commands. This
+   protects the sensitive information contained in the output of these
+   commands. For a full list of commands the driver can't monitor,
+   see the 
+   :spec:`MongoDB command monitoring specification </command-monitoring/command-monitoring.rst#security>`.
+
+Example
+^^^^^^^
+
+This example shows how to make a timer for database commands. The timer makes it so
+that every time the driver executes a command on your MongoDB instance
+you see a message letting you know how long it took to execute the command.
+
+To make a timer, do the following:
+
+#. Make a class with timer functionality that implements the ``CommandListener`` interface.
+#. Add an instance of the new class that implements ``CommandListener`` to a ``MongoClientSettings`` object.
+#. Configure a ``MongoClient`` instance with the ``MongoClientSettings`` object.
+
+The following code defines the ``CommandTimer`` class which implements the
```

Comment:
```suggestion
The following code defines the ``CommandTimer`` class, which implements the
```
Also applies to the other 2 event examples below.

Reply:
I believe the suggested edit should not be performed. I think "which implements the ____ interface" is a restrictive clause. I believe you only [add a comma before which for nonrestrictive clauses. ](https://www.grammarly.com/blog/comma-before-which/?&utm_source=google&utm_medium=cpc&utm_campaign=10273012991&utm_targetid=dsa-1233402314764&gclid=Cj0KCQjw2NyFBhDoARIsAMtHtZ7H64TbNm032HTN7BNzgoBZlmVoB8hKBXbACGNcXt6xhoHi3iiAg2saAp43EALw_wcB&gclsrc=aw.ds)